app/vminsert/netstorage/insert_ctx.go

```diff
@@ -134,26 +154,30 @@ func (ctx *InsertCtx) WriteDataPointExt(at *auth.Token, storageNodeIdx int, metr
 // FlushBufs flushes ctx bufs to remote storage nodes.
 func (ctx *InsertCtx) FlushBufs() error {
 	var firstErr error
-	for i := range ctx.bufRowss {
-		br := &ctx.bufRowss[i]
-		if len(br.buf) == 0 {
-			continue
-		}
-		if err := br.pushTo(storageNodes[i]); err != nil && firstErr == nil {
-			firstErr = err
+	for rgID, rg := range replicationGroups {
```

Comment:
It is better from performance PoV looping over `ctx.groupBufRowss`, since this eliminates map lookup:

```go
for _, bufRowss := range ctx.groupBufRowss {
    ...
}
```